rfc: Terraform provider (#2613)
edgelesssys/constellation: rfc/terraform-provider.md

# Constellation Terraform Provider

> The Constellation Terraform Provider should allow it's user to manage the full lifecycle of a Constellation cluster -- namely creation, updates, and deletion -- via Terraform.

Reviewer: ```suggestion
The Constellation Terraform Provider allows its user to manage the full lifecycle of a Constellation cluster -- namely creation, updates, and deletion -- via Terraform.
```

- The User needs to receive some sort of Access Token to the Cluster (e.g. a Kubeconfig, whether long- or short-lived)
when applying a configuration containing the provider and the resources listed below.

## Terraform Configuration Layout for the Constellation Resource

This resembles an examplary configuration of a Constellation cluster through Terraform. While the exact naming and struct layout
of the individual components may change until the implementation, the structure of the overall configuration looks like this:

```hcl
terraform {
  required_providers {
    constellation = {
      source  = "tbd/constellation"
      version = "2.13.0"
    }
  }
}

provider "constellation" { }

resource "constellation_cluster" "foo" {
    uid = "bar"
    name = "baz"
    image = data.constellation_image.ref # or provide manually crafted values
    kubernetes_version = "v1.27.6"
    constellation_microservice_version = "lockstep" # or pinned to a specific version
    debug = false
    init_endpoint = "10.10.10.10" # should use public ip of LB resource, ideally also provisioned through TF

>     kubernetes_api_endpoint = "10.10.10.10" # should use public ip of LB resource, ideally also provisioned through TF

Reviewer 2: Are these the only pieces of information that we need from the cloud infrastructure setup?

Author: Correct.

Reviewer 2: Thanks, but I'm slightly confused now looking at https://github.com/edgelesssys/constellation/pull/2613#discussion_r1401835204. Resolving this here and let's continue this topic over there.

    extra_microservices = {
        csi_driver = true
        # + more
        # possiblly also constellation microservices with version and maybe service options,
        # which would make constellation_microservice_version obsolete.
        # exact API TBD
    }
    master_secret = "foo" # updating this would force recreation of the cluster

>     init_secret = "bar" # maybe derive from master_secret, updating this would force recreation of the cluster

Reviewer 2: These secrets end up in the Terraform state, right? I believe we should mention that in this RFC.

    network_config = {
        # TBD
        # should contain CIDRs for pod network, service cidr, node network... for Cilium
        # and in-cluster Kubernetes API endpoint, e.g. for Kubelets

>     }

Reviewer 2: Are users expected to fill this in, or can this be derived from the cloud provider name and the state of the cloud infrastructure?

Author: I don't know for sure yet, @3u13r may have more information here?

Reviewer 3: We can get everything from the terraform as we also do today.

Reviewer 2: But the values in this resource are the only thing that our provider gets to see, correct? Would we need to reference other Terraform resources, like

```
  network_config {
    flavor = "gcp"
    podCIDR = other_resource.vpc.cidr
    # ...
  }
```

Author: The latter, I think: https://github.com/edgelesssys/constellation/blob/f6c44ee5cc6f1d097c0b2096f6a8e289e2210730/terraform/infrastructure/gcp/outputs.tf#L35-L37

Reviewer 3: We don't have to reference the other resource directly but there should be a direct link between the output of the vpc and the input of the Constellation provider (as Markus mentioned).

Author: Is this something that needs to be mentioned in the RFC? IMO, it should be agnostic about where the value comes from (output of another resource or directly supplied)

Reviewer 2: Only if the goal is to replace the TBD with an explicit list, which I don't think is needed. Maybe we could mention that some fields here need to be populated by other resources' outputs, but maybe that is obvious.

    attestation = data.constellation_attestation.attestation # or provide manually crafted values
}

# constellation_cluster provides:

> # constellation_cluster.foo.kubeconfig

Reviewer: I think the kubeconfig would just be stored in the output after init

> # constellation_cluster.foo.kubeconfig

Author: This would be the case if we tried to support importing resources. I don't know if we need to or if we want to yet, but in that case, we would have to make the Kubeconfig an input or find another way to pass it when importing the resource. For the init-case, we can make it so that the resource is only stored on init, and just returned from the state on any other operation. We cannot trivially recover the Kubeconfig with the mastersecret.

> # constellation_cluster.foo.kubeconfig

Reviewer 2: Got it: the kubeconfig is persisted in TF state, because it is part of the output, and this is where the provider gets it from for upgrades etc.

# constellation_cluster.foo.owner_id
# constellation_cluster.foo.cluster_id

data "constellation_attestation" "foo" {
    attestation_variant = "GCP-SEV-ES"
    image_version = "v2.13.0" # or "lockstep"
    maa_url = "https://www.example.com" # optional, for Azure only
}

# constellation_attestation provides:
# data.constellation_attestation.foo.measurements
# data.constellation_attestation.foo.attestation

data "constellation_image" "foo" {
    image_version = "v2.21.0" # or "lockstep"
    attestation_variant = "GCP-SEV-ES"
    csp = "GCP"
    region = "us-central1" # optional, for AWS only
}

# constellation_image provides:
# constellation_image.foo.reference (CSP-specific image reference)
```

The `constellation_cluster` resource is the main resource implemented by the provider.

> It resembles a Constellation cluster with a specific configuration.

Author: >  What does resemble mean? Is it not the same as a CLI created cluster?

It is, but I intended to say that we have the Terraform resource declaration on one side, and the actual cloud resources & data on the other side.

> What is the behavior of recreate?

Recreate means destroy and create, performed in that order. (i.e. you get a new cluster)

> Is it possible to perform a new init on an initialized cluster?

No

> It resembles a Constellation cluster with a specific configuration.

Reviewer: How would we enforce that the infra is recreated when the master secret changes? Atm, the infra is only related to the secret through a tag (only checked AWS). I don't think this would force a recreate. So I understand that we would design the infra module to force a recreate when it changes?

> It resembles a Constellation cluster with a specific configuration.

Reviewer: > It is, but I intended to say that we have the Terraform resource declaration on one side, and the actual cloud resources & data on the other side.

I would reword the sentence then. "The resource declares the provisioning / management of a Constellation cluster on the provided infrastructure."

> It resembles a Constellation cluster with a specific configuration.

Author: > How would we enforce that the infra is recreated when the master secret changes?

I think we should throw an error [like so](https://developer.hashicorp.com/terraform/plugin/framework/migrating/attributes-blocks/force-new#framework). Changing the underlying infrastructure is not necessary, the disks of the machines can just be overwritten with new images, leaving them in a "fresh" state. The only thing we need to warn the user about are previously created persistent volumes, which are encrypted with a secret derived from the old mastersecret and thus cannot be accessed with the new mastersecret.

> It resembles a Constellation cluster with a specific configuration.

Reviewer: Pls add this to the RFC.

> Applying it will create the cluster if not existing *or* if the required updates can't be performed in-place (e.g. changing the master secret),

Reviewer 4: We use the CRUD operations of terraform and the terraform state:

- When starting out, without terraform state, we expect the cluster to be uninitialized, so we perform the init
- The result of init is saved as terraform state
- Terraform will now call the update hook instead, so we know that we don't need to intialize the cluster again

> updates it with the according configuration if already existing, or deletes it if not present in the configuration but in the state.

Author: I don't understand the question. What do you mean exactly? The "deletes it if not" means that if the cluster is existing (in the Terraform state), but not declared in the config, it will be destroyed. (i.e. the normal Terraform behaviour, diffing state against declarations)

The "constellation_attestation" and "constellation_image" objects are [data sources](https://developer.hashicorp.com/terraform/language/data-sources),
which are objects that should be evaluated by the Provider each time the state is refreshed (i.e. each time any Terraform command that evaluates configuration against state),
but have no observable side effects. For image and attestation, this is required as the provider need to evaluate `latest` values or map CSP-agnostic image references (e.g. `v2.13.0`)

> to CSP-specific image references (e.g. `/CommunityGalleries/.../Image` for Azure). This is implemented as an idempotent API call and thus has no observable side-effects, but needs

Author: These are idempotent. We host an API that transforms CSP-agnostic image references (e.g. `v2.13.0`, `ref/main/stream/nightly/v2.13.0-pre.0.20231019162807-6c0a3b8efa49`) into CSP-specific image references (e.g. AMI-ID for AWS, Community Gallery ID for Azure, ...)

# Constellation Terraform Provider

The Constellation Terraform Provider allows its user to manage the full lifecycle of a Constellation cluster -- namely creation, updates, and deletion -- via Terraform.

## Design Goals

- The User needs to be able to perform Creation, Upgrades and Deletion of a Cluster through the provider.

> - Infrastructure provisioning should explicitly **not** be performed by the Constellation Terraform provider.

Author: Creation is better worded as "initialization" here, I guess.

- The User needs to receive some sort of Access Token to the Cluster (e.g. a Kubeconfig, whether long- or short-lived)
when applying a configuration containing the provider and the resources listed below.

> 

Reviewer: How would we include `status` information such as image / k8s upgrades? The user could manually fetch these with the CLI for troubleshooting cases? Same for `recovery`?

Author: The goal of the provider is to provide management and provisioning (i.e. create, update, delete) capabilities imo, not interaction with the cluster itself, like the CLI does. I don't think we need to achieve feature-parity with the CLI. The Terraform model would also not make it very easy to implement such without some hacks.

Reviewer: I agree

of the individual components may change until the implementation, the structure of the overall configuration looks like this:

```hcl
terraform {
  required_providers {
    constellation = {
      source  = "tbd/constellation"
      version = "2.13.0"
    }
  }
}

provider "constellation" { }

resource "constellation_cluster" "foo" {
    uid = "bar"
    name = "baz"
    image = data.constellation_image.ref # or provide manually crafted values
    kubernetes_version = "v1.27.6"
    constellation_microservice_version = "lockstep" # or pinned to a specific version
    debug = false
    init_endpoint = "10.10.10.10" # should use public ip of LB resource, ideally also provisioned through TF
    kubernetes_api_endpoint = "10.10.10.10" # should use public ip of LB resource, ideally also provisioned through TF
    extra_microservices = {
        csi_driver = true
        # + more
        # possiblly also constellation microservices with version and maybe service options,
        # which would make constellation_microservice_version obsolete.
        # exact API TBD
    }
    master_secret = "foo" # updating this would force recreation of the cluster

>     init_secret = "bar" # maybe derive from master_secret, updating this would force recreation of the cluster

Reviewer 5: Do I understand correctly that the user generates these secrets and pastes them into this TF config?
If so, is it common to have such secrets in a TF config? Or should they come from input variables, env variables, etc.?

Author: You could use the Terraform `random` provider (what we currently do) to generate this, load them from your HashiCorp vault by using the corresponding provider, etc.

I don't think you would expose them in the configuration files usually, but only in the state, which can be stored remotely, etc. and only have a reference to it, e.g. by using the `random` provider, in the configuration files - But the RFC should be agnostic of where the value comes from actually. When documenting our provider we could mention some recommendations of how secrets can be provisioned.
